crates/burn-tensor/src/tensor/activation/base.rs

```diff
@@ -125,13 +125,38 @@ pub fn sigmoid<const D: usize, B: Backend>(tensor: Tensor<B, D>) -> Tensor<B, D>
 
 /// Applies the log sigmoid function.
 pub fn log_sigmoid<const D: usize, B: Backend>(tensor: Tensor<B, D>) -> Tensor<B, D> {
+    /// To avoid overflow, we use the log-sum-exp trick.
+    ///
+    /// ```ignore
+    /// log(sigmoid(x)) = log(1/(1 + exp(-x)))
+    ///                 = log(1) - log(1 + exp(-x))
+    ///                 = -log(1 + exp(-x))
+    ///                 = -log(exp(0) + exp(-x))
+    /// ```
+    /// The `exp(t)` of even a moderate-magnitude positive number can be astronomically huge, so we
+    /// subtract the `max(t, 0)` of each value (where `t = -x` in this case). This results in the
+    /// following equivalence:
+    /// ```ignore
+    /// log(sigmoid(x)) = -(max(-x, 0) + log(exp(-max(-x, 0)) + exp(-x - max(-x, 0))))
+    /// ```
+    ///
+    /// This extends the range of values for which we obtain accurate results.
+    fn numerically_stable_log_sigmoid<const D: usize, B: Backend>(x: Tensor<B, D>) -> Tensor<B, D> {
+        // max(-x, 0)
+        let max_elem = x.clone().neg().max_pair(x.zeros_like());
+
+        // log(exp(-max(-x, 0)) + exp(-x - max(-x, 0)))
+        let z = (max_elem.clone().neg().exp() + (x.neg() - max_elem.clone()).exp()).log();
+
+        z.neg() - max_elem
+    }
     match B::FloatElem::precision() {
         Precision::Half => {
             let tensor_full = tensor.into_full_precision();
-            let tensor_tmp = tensor_full.neg().exp().add_scalar(1.0_f32).log().neg();
+            let tensor_tmp = numerically_stable_log_sigmoid(tensor_full);
             Tensor::from_full_precision(tensor_tmp)
         }
-        _ => tensor.neg().exp().add_scalar(1.0_f32).log().neg(),
+        _ => numerically_stable_log_sigmoid(tensor),
     }
 }
```

Comment:
I think a descent speedup for backends that don't implement fusion would be to move `log_sigmoid` and `sigmoid` into `burn_tensor::ops::activation` with the default implementation provided. We could then override those activations in backends that don't support fusion such as tch and candle.

Reply:
Agreed, I just tackled the scope of the current `log_sigmoid` implementation but that definitely came to mind.

Btw `sigmoid` is already in `ActivationOps` just not `log_sigmoid` yet.

Should I tackle this in a new PR or expand this one?